lxd/storage/drivers/drivers_lvm_utils.go

```diff
@@ -0,0 +1,693 @@
+
+	return nil
+}
+
+// countLogicalVolumes gets the count of volumes (both normal and thin) in a volume group.
+func (d *lvm) countLogicalVolumes(vgName string) (int, error) {
+	output, err := shared.RunCommand("vgs", "--noheadings", "-o", "lv_count", vgName)
+	if err != nil {
+		if d.isLVMNotFoundExitError(err) {
+			return -1, errLVMNotFound
+		}
+
+		return -1, err
+	}
+
+	output = strings.TrimSpace(output)
+	return strconv.Atoi(output)
+}
+
+// countThinVolumes gets the count of thin volumes in a thin pool.
+func (d *lvm) countThinVolumes(vgName, poolName string) (int, error) {
+	output, err := shared.RunCommand("lvs", "--noheadings", "-o", "thin_count", fmt.Sprintf("%s/%s", vgName, poolName))
+	if err != nil {
+		if d.isLVMNotFoundExitError(err) {
+			return -1, errLVMNotFound
+		}
+
+		return -1, err
+	}
+
+	output = strings.TrimSpace(output)
+	return strconv.Atoi(output)
+}
+
+// thinpoolExists checks whether the specified thinpool exists in a volume group.
+func (d *lvm) thinpoolExists(vgName string, poolName string) (bool, error) {
+	output, err := shared.RunCommand("lvs", "--noheadings", "-o", "lv_attr", fmt.Sprintf("%s/%s", vgName, poolName))
+	if err != nil {
+		if d.isLVMNotFoundExitError(err) {
+			return false, nil
+		}
+
+		return false, fmt.Errorf("Error checking for LVM thin pool \"%s\"", poolName)
+	}
+
+	// Found LV named poolname, check type:
+	attrs := strings.TrimSpace(string(output[:]))
+	if strings.HasPrefix(attrs, "t") {
+		return true, nil
+	}
+
+	return false, fmt.Errorf("LVM volume named \"%s\" exists but is not a thin pool", poolName)
+}
+
+// logicalVolumeExists checks whether the specified logical volume exists.
+func (d *lvm) logicalVolumeExists(volDevPath string) (bool, error) {
+	_, err := shared.RunCommand("lvs", "--noheadings", "-o", "lv_name", volDevPath)
+	if err != nil {
+		if d.isLVMNotFoundExitError(err) {
+			return false, nil
+		}
+
+		return false, fmt.Errorf("Error checking for LVM logical volume \"%s\"", volDevPath)
+	}
+
+	return true, nil
+}
+
+// createThinpool creates a thin pool logical volume.
+func (d *lvm) createThinpool(lvmVersion string, vgName string, thinPoolName string) error {
+	exists, err := d.thinpoolExists(vgName, thinPoolName)
+	if err != nil {
+		return err
+	}
+
+	if exists {
+		return nil
+	}
+
+	err = d.createDefaultThinPool(lvmVersion, vgName, thinPoolName)
+	if err != nil {
+		return err
+	}
+
+	poolExists, err := d.thinpoolExists(vgName, thinPoolName)
+	if err != nil {
+		return fmt.Errorf("Error checking for LVM thin pool \"%s\" in \"%s\": %v", thinPoolName, vgName, err)
+	}
+
+	if !poolExists {
+		return fmt.Errorf("LVM thin pool \"'%s\" does not exist in Volume Group \"%s\"", thinPoolName, vgName)
+	}
+
+	return nil
+}
+
+// createDefaultThinPool creates the default thinpool as 100% the size of the volume group with a 1G
+// meta data volume.
+func (d *lvm) createDefaultThinPool(lvmVersion, vgName, thinPoolName string) error {
```

Comment:
Only has a single caller, any reason to keep it separate?

Reply:
@stgraber for this function, I didn't to merge it into `Create` function, as that function is already rather long and complex, and this seems like a good logical unit to break it up (this function itself is also non-trivial and deals with different ways of creating a thin pool based on different versions of LVM).